### PR TITLE
Add extension on String to detect Emoji

### PR DESCRIPTION
**Pull request by daehn**
# What's in this PR?

Add an extension on String to check for Emoji, this is needed in order to disallow the input of emoji as username, as a check like
```swift
let allowed = CharacterSet(charactersIn: "abcdefghijklmnopqrstuvwxyz_").union(.decimalDigits)
CharacterSet(charactersIn: handle).isSubset(of: allowed)
```
will not catch emoji.

### CODE REVIEW

File: Tests/String+EmojiTests.swift [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+
+class String_EmojiTests: XCTestCase {
+
+    func testThatItDetectsAnEmoji() {
+        XCTAssertTrue("🍔".containsEmoji)
+    }
+
+    func testThatItDetectsMultipleEmoji() {
+        XCTAssertTrue("🍔😜🌮🎉🍕".containsEmoji)
```
Comment (marcoconti83) [inline Tests/String+EmojiTests.swift]: What about mixing text and emoji? Bonus points for mixing in non-ASCII text (arabic, ...)

Comment (marcoconti83) [inline Tests/String+EmojiTests.swift]: Ops, did not see the following test :)

Comment (daehn) [inline Tests/String+EmojiTests.swift]: There is one for mixing text and emoji, will add one for non-ASCII!

File: Source/Public/String+Emoji.swift [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+
+
+public extension String {
+
+    var containsEmoji: Bool {
+        if #available(iOS 9, *) {
+            return applyingTransform(.toUnicodeName, reverse: false) != self
```
Comment (typfel) [inline Source/Public/String+Emoji.swift]: Can we add little explanation why this detects emojis?

Comment (typfel) [inline Source/Public/String+Emoji.swift]: Can we add little explanation why this detects emojis?

**Comment by daehn** [comment]
This fails for non ASCII chars as they will also be transliterated into their unicode name, I will make this a throwing function and rename it so that it is clear it only works for latin characters.

**Comment by daehn** [comment]
The check if the string is composed from ascii characters is not feasible as it will also fire if there are emoji in the string. I wanted to make this more generic so it can be reused in other places, but as this would require more work I will close this and move the check into the valid handle check on the UI where we are already sure the string does not contain non ascii characters.

**Comment by daehn** [comment]
@marcoconti83 good catch!

File: Tests/String+EmojiTests.swift [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    func testThatItDoesNotDetectAnEmojiIfThereIsNone_EmptyString() {
+        XCTAssertFalse("".containsEmoji)
+    }
+
+    func testThatNonLatinsNotDetectedAsContainingEmoji() {
+        XCTAssertFalse("الأشخاص المفضلين".containsEmoji)
+    }
+
+    func testThatNonLatinWithEmojiIsDetectedAsContainingEmoji() {
+        XCTAssertTrue("الأشخاص المفضلين🙈".containsEmoji)
```
Comment (daehn) [inline Tests/String+EmojiTests.swift]: Maybe some more non-latin tests should be added

**Comment by daehn** [comment]
This can be reused on the UI to detect Emoji in order to display them in 3x as well.
I updated the PR to cover non-latin cases as well.

**Comment by daehn** [comment]
cc @marcoconti83 @typfel

**Pull request merged**